Firestore/core/src/firebase/firestore/nanopb/reader.h

```diff
@@ -80,9 +80,17 @@ class Reader {
    *
    * Call this method when reading a nested message. Provide a function to read
    * the message itself.
+   *
+   * @param invalid Value to return when an error occurs.
    */
   template <typename T>
-  T ReadNestedMessage(const std::function<T(Reader*)>& read_message_fn);
+  T ReadNestedMessage(T invalid,
```

Comment:
I'm all in favor of naming parallelism so I think the name is fine, but this seems like a strange interface.

Largely the contract of these methods is that they have to check whether or not the reader is ok before doing anything useful. Why not just impose this requirement on implementations of the read_message_fn?

The functions in this PR already have to deal with a failure to read, in which case they already know how to create an invalid instance. Why avoid calling into that function?

I could see `ReadNestedMessage` doing something more like this:

```c++
if (!status_.ok()) {
  return read_message_fn(this);
}

pb_istream_t raw_substream;
if (!pb_make_string_substream(&stream_, &raw_substream)) {
  status_ = util::Status(FirestoreErrorCode::DataLoss, PB_GET_ERROR(&stream_));
  pb_close_string_substream(&stream_, &raw_substream);
  return read_message_fn(this);
}

// otherwise handle real substream as before
```